README.md
# JSON-RPC 2.0 Client

Copyright (c) Vladimir Dzhuvinov, 2010 - 2021

Provides a client-side class for dispatching requests and notifications to a 
JSON-RPC 2.0 server. The JSON-RPC 2.0 messages are sent by HTTP(S) POST.

Optional client session features:

* Customise the "Content-Type" header in HTTP POST requests.

* Set an "Origin" header in HTTP POST requests to simulate Cross-Origin 
  Resource Sharing (CORS) requests from a browser.
  
* Accept HTTP cookies (if client sessions are established by this mean instead 
  of through the JSON-RPC protocol itself).
  
* Customise the allowable "Content-Type" header values in HTTP POST responses.
 
* Set an HTTP proxy.

* Enable HTTP response compression using GZIP and DEFLATE content encoding.
  
* Preserve parse order of JSON object members in JSON-RPC 2.0 response results 
  (for human facing clients, e.g. the JSON-RPC 2.0 Shell).
  
* Ignore version 2.0 checks when parsing responses to allow client sessions to 
  older JSON-RPC (1.0) servers.

* Parse non-standard attributes appended in JSON-RPC 2.0 responses.

* ~Trust all X.509 server certificates (for HTTPS connections), including
  self-signed.~ (**Removed to avoid triggering X509TrustManager's security 
  vulnerabilities not compliant with Google Play Policies in Google Play Market.
  Details: <https://support.google.com/faqs/answer/6346016>**)


Requirements:

* Java 1.6 or later

* The package depends on the JSON-RPC 2.0 Base library and the JSON Smart
  library for JSON encoding and decoding (fork of the popular JSON.simple
  toolkit, but with more efficient parsing).


Visit the library home page for usage, examples and updates:

<http://software.dzhuvinov.com/json-rpc-2.0-client.html>



CHANGELOG.txt
CHANGE LOG

JsonRpc2-Client, Copyright (c) Vladimir Dzhuvinov, 2010 - 2021


version 1.0 (2010-11-09)
	* First release.

version 1.1 (2011-03-10)
	* Major refactoring and API change.
	* Adds optional setting for custom "Content-Type" header in HTTP POST
	  requests.
	* Adds optional setting to customise the allowable "Content-Type" header
	  values in HTTP POST responses.
	* Adds optional setting to disable strict JSON-RPC 2.0 parsing of
	  responses to allow sessions with older JSON-RPC (1.0) servers.
	* Adds optional setting to trust self-signed server sertificates for
	  secure HTTPS connections.

version 1.2 (2011-03-29)
	* Skips ID match checking for JSON-RPC 2.0 responses indicating errors
	  -32700 (parse error), -32600 (invalid request) and -32603 (internal
	  error).

version 1.3 (2011-07-11)
	* Raises minimal JSON-RPC 2.0 Base requirement to version 1.19.
	* Deprecates JSONRPC2Session.disableStrictParsing() in favour of
	  JSONRPC2Session.ignoreVersion().
	* Deprecates JSONRPC2Session.strictParsingDisabled() in favour of
	  JSONRPC2Session.ignoresVersion().
	* Adds option to parse non-standard attributes appended to 
	  JSON-RPC 2.0 responses.

version 1.4 (2011-07-13)
	* Moves all optional session properties and methods into a separate
	  JSONPRC2SessionOptions class.
	* Adds a JUnit test for the JSONRPC2SessionOptions class.

version 1.5 (2011-07-18)
	* Adds an interface for specifying a custom HTTP URL connection
	  configurator.

version 1.6 (2011-08-23)
	* Adds an interface and class for inspecting the raw HTTP response.
	* Adds a configuration option to accept HTTP cookies.

version 1.7 (2012-02-16)
	* Provides additional JSON-RPC 2.0 session exception cause types.

version 1.7.1 (2012-04-03)
	* Updates JSON-RPC 2.0 Base JAR to 1.25.1 (JSON Smart 1.1.1).

version 1.7.2 (2012-07-14)
	* Updates JSON-RPC 2.0 Base JAR to 1.27.

version 1.8 (2012-08-26)
	* Introduces option for controlling HTTP connection and read timeouts.

version 1.9 (2012-12-01)
	* Upgrades JSON-RPC 2.0 Base JAR to 1.30.

version 1.10 (2012-12-13)
	* Adds HTTP proxy support.

version 1.11 (2012-12-28)
	* Fixes JSONRPC2SessionException message on missing Content-Type 
	  header in HTTP response.
	* Upgrades JSON-RPC 2.0 Base JAR to 1.31.
	
version 1.12 (2013-01-22)
	* Adds support for HTTP response compression using GZIP and DEFLATE 
	  content encoding.
	* Switches cookie handling to java.net.CookieManager to add support for
	  cookie replacement and expiration.

version 1.13 (2013-03-18)
	* Forces UTF-8 output for POSTed JSON-RPC 2.0 strings.

version 1.14 (2013-03-29)
	* Switches project build from Ant to Maven.
	* Expects UTF-8 input for HTTP responses.
	* Upgrades to JSON-RPC 2.0 Base 1.34.

version 1.14.1 (2013-03-30)
	* Updates Maven pom.xml.
	
version 1.14.2 (2013-03-30)
	* Fixes bug in assembly-dist.xml.

version 1.14.3 (2013-04-02)
	* Publishes library to Maven Central.

version 1.14.4 (2013-06-13)
	* Fixes exception reporting on request / response ID mismatch.

version 1.15 (2015-03-16)
    * Cleans up code.
    * Upgrades to JSON-RPC 2.0 Base 1.36.
    * Upgrades to JSON Smart 1.3.1.

version 1.16 (2016-02-13)
    * Closes the URL connection I/O streams after use.

version 1.16.1 (2016-02-13)
    * Closes the URL connection I/O streams explicitly upon encountering a
      missing trust-all socket factory for HTTPS connections.

version 1.16.2 (2016-02-18)
    * Closes the error stream of the HTTP(S) URL connection after use.

version 1.16.3 (2016-02-18)
    * Performs a null check before closing the input, output and error streams
      of the HTTP(S) URL connection after use.

version 1.16.4 (2016-02-18)
    * Refactors closing of the input, output and error streams of the HTTP(S)
      URL connection after use.

version 1.16.5 (2021-03-16)
    * Updates to JSON-RPC 2.0 Base 1.38.2.

version 1.16.6 (2021-04-12)
    * Removed optional setting to trust self-signed server sertificates for
      secure HTTPS connections, to avoid triggering X509TrustManager's security
      vulnerabilities not compliant with Google Play Policies in Google Play Market.
      Details: https://support.google.com/faqs/answer/6346016

src/test/java/com/thetransactioncompany/jsonrpc2/client/Test.java
package com.thetransactioncompany.jsonrpc2.client;


import java.net.MalformedURLException;
import java.net.URL;
import java.util.Arrays;

import com.thetransactioncompany.jsonrpc2.JSONRPC2Notification;
import com.thetransactioncompany.jsonrpc2.JSONRPC2Request;
import com.thetransactioncompany.jsonrpc2.JSONRPC2Response;
import junit.framework.TestCase;
import net.minidev.json.JSONObject;


/**
 * Tests the JSONRPC2Session class.
 *
 * @author Vladimir Dzhuvinov
 */
public class Test extends TestCase {

	public static final String URL_HTTP_GOOD = "https://demo.c2id.com/json2ldap/";
	public static final String URL_HTTP_BAD_HOST = "http://bad-host.com/ws/";
	public static final String URL_HTTP_BAD_PORT = "http://localhost:9999/";
	public static final String URL_HTTP_BAD_IP = "http://192.168.0.99";
	public static final String URL_HTTP_BAD_SERVICE = "http://google.com/";
	public static final String URL_HTTPS_SELF_SIGNED = "https://shakespeare.dzhuvinov.com:8443/json2ldap/";

	public static final String RPC_METHOD_GOOD = "ws.getTime";

	
	public Test(String name) {

		super(name);
	}


	public void testRequest() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 0);

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);

		} catch (JSONRPC2SessionException e) {
			fail(e.getMessage());
		}

		System.out.println("Response: " + response);
	}


	// Disable temporarily
	public void _testRequestEnableCompression() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 0);

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);

		session.getOptions().enableCompression(true);
		assertTrue(session.getOptions().enableCompression());

		ResponseInspector responseInspector = new ResponseInspector();
		session.setRawResponseInspector(responseInspector);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);

		} catch (JSONRPC2SessionException e) {
			fail(e.getMessage());
		}

		RawResponse rawResponse = responseInspector.getLastRawResponse();

		assertEquals("gzip", rawResponse.getContentEncoding());

		System.out.println("Response (decompressed): " + response);
	}


	public void testNotification() {

		// Assumes a Json2Ldap service

		JSONRPC2Notification notification = new JSONRPC2Notification(RPC_METHOD_GOOD);

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		try {
			session.send(notification);

		} catch (JSONRPC2SessionException e) {
			fail(e.getMessage());
		}
	}


	public void testNetworkException1() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 0);

		URL url = null;

		try {
			url = new URL(URL_HTTP_BAD_HOST);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);
		session.getOptions().setConnectTimeout(1000); // 1 second

		JSONRPC2Response response = null;

		try {
			response = session.send(request);
			fail("Failed to raise network exception");

		} catch (JSONRPC2SessionException e) {
			// ok
			assertEquals(JSONRPC2SessionException.NETWORK_EXCEPTION, e.getCauseType());
			System.out.println(e.getMessage() + " [" + e.getCause().getClass().getName() + "]");
		}
	}


	public void testNetworkException2() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 0);

		URL url = null;

		try {
			url = new URL(URL_HTTP_BAD_PORT);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);
		session.getOptions().setConnectTimeout(1000); // 1 second

		JSONRPC2Response response = null;

		try {
			response = session.send(request);
			fail("Failed to raise network exception");

		} catch (JSONRPC2SessionException e) {
			// ok
			assertEquals(JSONRPC2SessionException.NETWORK_EXCEPTION, e.getCauseType());
			System.out.println(e.getMessage() + " [" + e.getCause().getClass().getName() + "]");
		}
	}



	public void testNetworkException3() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 0);

		URL url = null;

		try {
			url = new URL(URL_HTTP_BAD_IP);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);
		session.getOptions().setConnectTimeout(1000); // 1 second

		JSONRPC2Response response = null;

		try {
			response = session.send(request);
			fail("Failed to raise network exception");

		} catch (JSONRPC2SessionException e) {
			// ok
			assertEquals(JSONRPC2SessionException.NETWORK_EXCEPTION, e.getCauseType());
			System.out.println(e.getMessage() + " [" + e.getCause().getClass().getName() + "]");
		}
	}


	public void testNetworkException4() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 0);

		URL url = null;

		try {
			// NO POST (HTTP 405)
			url = new URL(URL_HTTP_BAD_SERVICE);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);
			fail("Failed to raise JSON-RPC 2.0 exception");

		} catch (JSONRPC2SessionException e) {
			// ok
			System.out.println(e.getMessage() + " [" + e.getCause().getClass().getName() + "]");

			assertEquals(JSONRPC2SessionException.NETWORK_EXCEPTION, e.getCauseType());
		}
	}


	public void testRequestIDMatchingBoolean() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, true);

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);

		} catch (JSONRPC2SessionException e) {

			fail(e.getMessage());
		}

		System.out.println("Response: " + response);
	}


	public void testRequestIDMatchingNumber() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, 3.14);

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);

		} catch (JSONRPC2SessionException e) {

			fail(e.getMessage());
		}

		System.out.println("Response: " + response);
	}


	public void testRequestIDMatchingString() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, "abc");

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);

		} catch (JSONRPC2SessionException e) {

			fail(e.getMessage());
		}

		System.out.println("Response: " + response);
	}


	public void testRequestIDMatchingNull() {

		// Assumes a Json2Ldap service

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, null);

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);


		JSONRPC2Response response = null;

		try {
			response = session.send(request);

		} catch (JSONRPC2SessionException e) {

			fail(e.getMessage());
		}

		System.out.println("Response: " + response);
	}

	public void testRequestContentType() {

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);

		session.getOptions().setRequestContentType("application/json+rpc");

		assertEquals("application/json+rpc", session.getOptions().getRequestContentType());
	}


	public void testResponseContentType() {

		URL url = null;

		try {
			url = new URL(URL_HTTP_GOOD);

		} catch (MalformedURLException e) {
			fail(e.getMessage());
		}

		JSONRPC2Session session = new JSONRPC2Session(url);

		session.getOptions().setAllowedResponseContentTypes(new String[]{"application/json+rpc"});

		JSONRPC2Request request = new JSONRPC2Request(RPC_METHOD_GOOD, null);

		try {
			session.send(request);

			fail("Failed to raise unexpected content-type exception");
		
		} catch (JSONRPC2SessionException e) {

			System.out.println(e.getMessage());
		}
	}


	public void _testCyrillic()
		throws Exception {

		JSONRPC2Session session = new JSONRPC2Session(new URL("http://localhost:8080/json2ldap/"));
		session.getOptions().setRequestContentType("application/json;charset=utf8");

		JSONObject params = new JSONObject();
		params.put("host", "localhost");
		params.put("port", 1389);
		JSONRPC2Response response = session.send(new JSONRPC2Request("ldap.connect", params, "1"));

		System.out.println(response);

		String cid = (String)((JSONObject)response.getResult()).get("CID");

		System.out.println("CID: " + cid);

		params = new JSONObject();
		params.put("CID", cid);
		params.put("DN", "cn=Directory Manager");
		params.put("password", "secret");

		System.out.println(session.send(new JSONRPC2Request("ldap.simpleBind", params, "3")));

		params = new JSONObject();
		params.put("CID", cid);
		params.put("baseDN", "ou=people,dc=wonderland,dc=net");
		params.put("filter", "(givenName=Алиса)");
		params.put("scope", "ONE");
		params.put("attributes", Arrays.asList("givenName", "uid"));

		System.out.println(session.send(new JSONRPC2Request("ldap.search", params, "4")));

		params = new JSONObject();
		params.put("CID", cid);
		session.send(new JSONRPC2Request("ldap.close", params, "2"));
	}
}


src/main/java/com/thetransactioncompany/jsonrpc2/client/JSONRPC2SessionOptions.java
package com.thetransactioncompany.jsonrpc2.client;


import java.net.Proxy;

import java.util.regex.Pattern;


/**
 * Optional settings for JSON-RPC 2.0 client sessions. The no-argument 
 * constructor specifies the default settings that the {@link JSONRPC2Session}
 * uses. To apply different ones instantiate a new settings instance, set the
 * desired ones to your liking, and then {@link JSONRPC2Session#setOptions pass} 
 * it to your {@link JSONRPC2Session} instance.
 * 
 * <p>Overview of the available session options:
 *
 * <ul>
 *     <li>Customise the "Content-Type" header in HTTP POST requests.
 *     <li>Set an "Origin" header in HTTP POST requests to simulate 
 *         Cross-Origin Resource Sharing (CORS) requests from a browser.
 *     <li>Accept HTTP cookies (if client sessions are established by this 
 *         mean instead of through the JSON-RPC protocol itself).
 *     <li>Customise the allowable "Content-Type" header values in HTTP POST
 *         responses.
 *     <li>Preserve parse order of JSON object members in JSON-RPC 2.0 response
 *         results (for human-facing clients, e.g. the JSON-RPC 2.0 Shell).
 *     <li>Ignore version 2.0 checks when parsing responses to allow client 
 *         sessions to older JSON-RPC (1.0) servers.
 *     <li>Parse non-standard attributes in JSON-RPC 2.0 responses.
 *     <li>Set an HTTP connect timeout.
 *     <li>Set an HTTP read timeout.
 *     <li>Set an HTTP proxy.
 *     <li>Enable HTTP response compression (using GZIP or DEFLATE content
 *         encoding).
 * </ul>
 *
 * @since 1.4
 * @author Vladimir Dzhuvinov
 */
public class JSONRPC2SessionOptions {
	
	
	/**
	 * The "Content-Type" (MIME) header value of HTTP POST requests. If
	 * {@code null} the header will not be set.
	 */
	private String requestContentType = DEFAULT_CONTENT_TYPE;
	
	
	/**
	 * The default "Content-Type" (MIME) header value of HTTP POST 
	 * requests. Set to {@code application/json}.
	 */
	public static final String DEFAULT_CONTENT_TYPE = "application/json";
	
	
	/**
	 * The allowed "Content-Type" (MIME) header values of HTTP responses. 
	 * If {@code null} any header value will be accepted.
	 */
	private String[] allowedResponseContentTypes = 
		DEFAULT_ALLOWED_RESPONSE_CONTENT_TYPES;
	
	
	/**
	 * The default allowed "Content-Type" (MIME) header values of HTTP
	 * responses. Set to {@code application/json} and {@code text/plain}.
	 */
	public static final String[] DEFAULT_ALLOWED_RESPONSE_CONTENT_TYPES =
		{"application/json", "text/plain"};
	 
	 
	/** 
	 * Optional CORS "Origin" header. If {@code null} the header will not 
	 * be set.
	 */
	private String origin = DEFAULT_ORIGIN;
	
	
	/**
	 * The default CORS "Origin" header value. Set to {@code null} (none).
	 */
	public static final String DEFAULT_ORIGIN = null;
	
	
	/**
	 * Specifies whether to accept HTTP cookies.
	 */
	private boolean acceptCookies = DEFAULT_ACCEPT_COOKIES;
	
	
	/**
	 * The default HTTP cookie acceptance policy.
	 */
	public static final boolean DEFAULT_ACCEPT_COOKIES = false;
	
	
	/**
	 * If {@code true} the order of parsed JSON object members must be
	 * preserved.
	 */
	private boolean preserveObjectMemberOrder = 
		DEFAULT_PRESERVE_OBJECT_MEMBER_ORDER;
	
	
	/**
	 * The default policy for preserving the order of parsed JSON object
	 * members. Set to {@code false} (no preserve).
	 */
	public static final boolean DEFAULT_PRESERVE_OBJECT_MEMBER_ORDER = false;
	
	
	/**
	 * If {@code true} version 2.0 checking of received responses must be
	 * disabled.
	 */
	private boolean ignoreVersion = DEFAULT_IGNORE_VERSION ;
	
	
	/**
	 * The default policy for version 2.0 checking. Set to {@code false}
	 * (strict checking).
	 */
	public static final boolean DEFAULT_IGNORE_VERSION = false;
	
	
	/**
	 * If {@code true} non-standard attributes appended to the JSON-RPC 2.0
	 * responses must be parsed too.
	 */
	private boolean parseNonStdAttributes = DEFAULT_PARSE_NON_STD_ATTRIBUTES;
	
	
	/**
	 * The default policy for parsing non-standard attributes in JSON-RPC 
	 * 2.0 messages. Set to {@code false} (non-standard attributes ignored).
	 */
	public static final boolean DEFAULT_PARSE_NON_STD_ATTRIBUTES = false;
	
	
	/**
	 * The HTTP connect timeout, in milliseconds. Zero implies the option 
	 * is disabled (timeout of infinity).
	 */
	private int connectTimeout = DEFAULT_CONNECT_TIMEOUT;
	
	
	/**
	 * The default HTTP connect timeout. Set to zero (disabled).
	 */
	public static final int DEFAULT_CONNECT_TIMEOUT = 0;
	
	
	/**
	 * The HTTP read timeout, in milliseconds. Zero implies the option is
	 * disabled (timeout of infinity).
	 */
	private int readTimeout = DEFAULT_READ_TIMEOUT;
	
	
	/**
	 * The default HTTP read timeout. Set to zero (disabled).
	 */
	public static final int DEFAULT_READ_TIMEOUT = 0;


	/**
	 * Optional HTTP proxy.
	 */
	private Proxy proxy = null;


	/**
	 * Enable / disable HTTP GZIP and DEFLATE compression.
	 */
	private boolean enableCompression = DEFAULT_ENABLE_COMPRESSION;


	/**
	 * The default HTTP GZIP and DEFLATE compression enable policy.
	 */
	public static final boolean DEFAULT_ENABLE_COMPRESSION = false;



	/**
	 * Creates a new default JSON-RPC 2.0 client session options instance.
	 *
	 * <p>The "Content-Type" (MIME) header value of HTTP POST requests will
	 * be set to "application/json". To change it use 
	 * {@link #setRequestContentType}.
	 *
	 * <p>"Origin" HTTP headers will not be added. To add one use
	 * {@link #setOrigin}.
	 * 
	 * <p>HTTP cookies will be ignored. To accept cookies, e.g. for 
	 * browser-like session handling, use {@link #acceptCookies}.
	 *
	 * <p>The allowed HTTP response content types are set to 
	 * "application/json" and "text/plain". To change them use
	 * {@link #setAllowedResponseContentTypes}.
	 *
	 * <p>The parse order of JSON object members in JSON-RPC 2.0 response
	 * results will not be preserved. To change this behaviour use
	 * {@link #preserveParseOrder}.
	 *
	 * <p>Strict 2.0 version checking will be performed. To ignore the
	 * JSON-RPC version attribute use {@link #ignoreVersion(boolean)}.
	 *
	 * <p>HTTP connect timeouts will be disabled. To specify a value use
	 * {@link #setConnectTimeout}.
	 *
	 * <p>HTTP read timeouts will be disabled. To specify a value use
	 * {@link #setReadTimeout}.
	 *
	 * <p>No proxy is used. To specify one use {@link #setProxy}.
	 *
	 * <p>HTTP response compression (GZIP or DEFLATE) is disabled. To 
	 * enable it use {@link #enableCompression(boolean)}.
	 */
	public JSONRPC2SessionOptions() {
	
		// check fields for default init values
	}
	
	
	/**
	 * Gets the value of the "Content-Type" (MIME) header for HTTP POST
	 * requests.
	 *
	 * @return The "Content-Type" (MIME) header value, {@code null} if the
	 *         header is not added to HTTP POST requests.
	 */
	public String getRequestContentType() {
	
		return requestContentType;
	}
	
	
	/**
	 * Sets the value of the HTTP "Content-Type" (MIME) header. Use this 
	 * method if you wish to change the default "application/json" content 
	 * type.
	 *
	 * @param contentType The value of the "Content-Type" (MIME) header
	 *                    in HTTP POST requests, {@code null} to suppress
	 *                    the header.
	 */
	public void setRequestContentType(final String contentType) {
	
		this.requestContentType = contentType;
	}
	
	
	/**
	 * Gets the value of the "Origin" HTTP header.
	 *
	 * <p>This header can be used to simulate Cross-Origin Resource Sharing
	 * (CORS) requests from a browser.
	 *
	 * @return The "Origin" header value, {@code null} if the header is not
	 *         added to HTTP requests.
	 */
	public String getOrigin() {
		
		return origin;
	}
	
	
	/**
	 * Sets the value of the "Origin" HTTP header.
	 *
	 * <p>This header can be used to simulate Cross-Origin Resource Sharing
	 * (CORS) requests from a browser.
	 *
	 * @param origin The value of the "Origin" header in HTTP requests, 
	 *               {@code null} to suppress the header.
	 */
	public void setOrigin(final String origin) {
	
		this.origin = origin;
	}
	
	
	/**
	 * Returns {@code true} if HTTP cookies are accepted, else 
	 * {@code false} if they are ignored.
	 *
	 * @return {@code true} if HTTP cookies are accepted, else 
	 *         {@code false}.
	 */
	public boolean acceptCookies() {
	
		return acceptCookies;
	}
	
	
	/**
	 * Specifies whether to accept HTTP cookies contained in the server 
	 * response. Some JSON-RPC servers may use cookies instead of tokens
	 * passed through the JSON-RPC protocol itself to establish client 
	 * sessions.
	 *
	 * @param acceptCookies {@code true} to accept HTTP cookies, else
	 *                      {@code false} to ignore them.
	 */
	public void acceptCookies(final boolean acceptCookies) {
	
		this.acceptCookies = acceptCookies;
	}
	
	
	/**
	 * Gets the allowed "Content-Type" (MIME) header values of HTTP 
	 * responses. 
	 *
	 * <p>The {@code JSONRPC2Session.send(...)} method will throw a
	 * {@link JSONRPC2SessionException#UNEXPECTED_CONTENT_TYPE} if the
	 * received HTTP response "Content-Type" (MIME) header value is not
	 * allowed.
	 *
	 * @return The allowed header values, if {@code null} any header value 
	 *         is allowed.
	 */
	public String[] getAllowedResponseContentTypes() {
	
		return allowedResponseContentTypes;
	}
	
	
	/**
	 * Sets the allowed "Content-Type" (MIME) header values of HTTP 
	 * responses.
	 *
	 * <p>The {@code JSONRPC2Session.send(...)} method will throw a 
	 * {@link JSONRPC2SessionException#UNEXPECTED_CONTENT_TYPE} if the
	 * received HTTP response "Content-Type" (MIME) header value is not
	 * allowed.
	 *
	 * @param contentTypes The allowed header values, {@code null} to allow
	 *                     any header value.
	 */
	public void setAllowedResponseContentTypes(final String[] contentTypes) {
	
		this.allowedResponseContentTypes = contentTypes;
	}
	
	
	/**
	 * Checks if the specified HTTP "Content-Type" (MIME) header value is 
	 * allowed.
	 *
	 * @param contentType The "Content-Type" (MIME) header value.
	 *
	 * @return {@code true} if the content type is allowed, else 
	 *         {@code false}.
	 */
	public boolean isAllowedResponseContentType(final String contentType) {
	
		// Allow any?
		if (allowedResponseContentTypes == null)
			return true;
		
		if (contentType == null)
			return false; // missing
		
		for (String t: allowedResponseContentTypes) {

			// Note: the content type may include optional parameters, 
			// which must be ignored during matching, e.g.
			// "application/json; charset=ISO-8859-1; ..."	
			if (contentType.matches("^" + Pattern.quote(t) + "(\\s|;|$)?.*"))
				return true;
		}
		
		return false; // nothing matched
	}
	
	
	/**
	 * Returns {@code true} if the member order of parsed JSON objects in
	 * JSON-RPC 2.0 response results is preserved.
	 *
	 * @return {@code true} if the parse order of JSON object members is
	 *         preserved, else {@code false}.
	 */
	public boolean preservesParseOrder() {
	
		return preserveObjectMemberOrder;
	}
	
	
	/**
	 * Controls the behaviour of the JSON parser when processing object
	 * members in JSON-RPC 2.0 response results. The default behaviour is
	 * to store the members in a {@code java.util.HashMap} in a 
	 * non-deterministic order. To preserve the original parse order pass a 
	 * boolean {@code true} to this method. Note that this will slow down 
	 * parsing and retrieval performance somewhat.
	 *
	 * @param preserve If {@code true} the parse order of JSON object 
	 *                 members will be preserved, else not.
	 */
	public void preserveParseOrder(final boolean preserve) {
	
		preserveObjectMemberOrder = preserve;
	}
	
	
	/**
	 * Returns {@code true} if strict parsing of received JSON-RPC 2.0
	 * responses is disabled and the "jsonrpc" version attribute is not 
	 * checked for "2.0" equality. Returns {@code false} if received 
	 * JSON-RPC 2.0 responses must strictly conform to the JSON-RPC 2.0 
	 * specification.
	 *
	 * @return {@code true} if the {@code "jsonrpc":"2.0"} version 
	 *         attribute is ignored, {@code false} if parsing is strict.
	 */
	public boolean ignoresVersion() {
	
		return ignoreVersion;
	}
	
	
	/**
	 * Controls the strictness of the JSON-RPC 2.0 response parser. The
	 * default behaviour is to check responses for strict compliance to
	 * the JSON-RPC 2.0 specification. By passing a boolean {@code true}
	 * parsing is relaxed and the "jsonrpc" version attribute will not be
	 * checked for "2.0" equality.
	 *
	 * @param ignore {@code true} to ignore the 2.0 {@code "jsonrpc":"2.0"} 
	 *               version attribute, {@code false} for strict parsing.
	 */
	public void ignoreVersion(final boolean ignore) {
	
		ignoreVersion = ignore;
	}
	
	
	/**
	 * Specifies whether to parse non-standard attributes found in JSON-RPC 
	 * 2.0 responses. 
	 * 
	 * @param enable {@code true} to parse non-standard attributes, else 
	 *               {@code false}.
	 */
	public void parseNonStdAttributes(final boolean enable) {
	
		parseNonStdAttributes = enable;
	}
	
	
	/**
	 * Returns {@code true} if non-standard attributes in JSON-RPC 2.0 
	 * responses are parsed. 
	 *
	 * @return {@code true} if non-standard attributes are parsed, else 
	 *         {@code false}.
	 */
	public boolean parsesNonStdAttributes() {
	
		return parseNonStdAttributes;
	}
	
	
	/**
	 * Sets the HTTP connect timeout.
	 *
	 * @since 1.8
	 *
	 * @param timeout The HTTP connect timeout, in milliseconds. Zero
	 *                implies the option is disabled (timeout of infinity).
	 */
	public void setConnectTimeout(final int timeout) {
	
		if (timeout < 0)
			throw new IllegalArgumentException("The HTTP connect timeout must be zero or positive");
		
		connectTimeout = timeout;
	}
	
	
	/**
	 * Gets the HTTP connect timeout.
	 *
	 * @since 1.8
	 *
	 * @return The HTTP connect timeout, in milliseconds. Zero implies the
	 *         option is disabled (timeout of infinity).
	 */
	public int getConnectTimeout() {
	
		return connectTimeout;
	}
	
	
	/**
	 * Sets the HTTP read timeout.
	 *
	 * @since 1.8
	 *
	 * @param timeout The HTTP read timeout, in milliseconds. Zero implies
	 *                the option is disabled (timeout of infinity).
	 */
	public void setReadTimeout(final int timeout) {
	
		if (timeout < 0)
			throw new IllegalArgumentException("The HTTP read timeout must be zero or positive");
		
		readTimeout = timeout;
	}
	
	
	/**
	 * Gets the HTTP read timeout.
	 *
	 * @since 1.8
	 *
	 * @return The HTTP read timeout, in milliseconds. Zero implies the 
	 *         option is disabled (timeout of infinity).
	 */
	public int getReadTimeout() {
	
		return readTimeout;
	}


	/**
	 * Sets an HTTP proxy.
	 *
	 * @since 1.10
	 *
	 * @param proxy The HTTP proxy to use, {@code null} if none.
	 */
	public void setProxy(final Proxy proxy) {

		this.proxy = proxy;
	}


	/**
	 * Gets the HTTP proxy.
	 *
	 * @since 1.10
	 *
	 * @return The HTTP proxy to use, {@code null} if none.
	 */
	public Proxy getProxy() {

		return proxy;
	}


	/**
	 * Enables or disables HTTP response compression using GZIP or DEFLATE
	 * content encoding. If compression is enabled but the HTTP server 
	 * doesn't support compression this setting will have no effect.
	 *
	 * @param enable If {@code true} HTTP compression will be enabled, 
	 *               else compression will be disabled.
	 */
	public void enableCompression(final boolean enable) {

		enableCompression = enable;
	}


	/**
	 * Checks if HTTP response compression using GZIP or DEFLATE content 
	 * encoding is enabled or disabled. If compression is enabled but the 
	 * HTTP server doesn't support compression this setting will have no 
	 * effect.
	 *
	 * @return {@code true} if HTTP compression is enabled, else 
	 *         {@code false}.
	 */
	public boolean enableCompression() {

		return enableCompression;
	}
	

}
